File: warehouse/wareaddr.js
var sw_grid = new dhtmlXGridObject({
                        parent: window.app.Panels["def"],       
                        auto_width: true,
                        columns:[
                                  
	  { type:"ch", sort:"str",  align:"left",  width:"20",  label:"Печать" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"ID1" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"status" },
	  
	  { type:"ro", sort:"str",  align:"left",  width:"100", label:["Адрес","#text_filter"] },
	  { type:"ro", sort:"str",  align:"left",  width:"56",  label:["Тип ячейки", "#select_filter"] },
	  { type:"ro", sort:"str",  align:"left",  width:"100", label:["Тип товара", "#select_filter"] },
	  { type:"ro", sort:"str",  align:"right", width:"80",  label:["Код товара","#text_filter"] },
	
	  { type:"ro", sort:"str",  align:"left",  width:"150", label:["Наименование","#text_filter"] },
	  { type:"ro", sort:"int",  align:"right", width:"50",  label:"В коробке" },
	  { type:"ro", sort:"int",  align:"right", width:"40",  label:"Коробок", style: "color: red;" },  
	  { type:"ro", sort:"int",  align:"right", width:"46",  label:"Кол-во" },
	  
	  { type:"ro", sort:"int",  align:"right", width:"46",  label:"Ожид. расход(в коробках)" },
	  
	  
	  { type:"ro", sort:"str",  align:"right", width:"50",  label:["Партия","#text_filter"] },
	  { type:"ro", sort:"date", align:"left",  width:"64",  label:"Годен до" },
	  { type:"ro", sort:"str",  align:"left",  width:"64",  label:["Статус","#text_filter"] }, 
	  { type:"ro", sort:"int",  align:"right", width:"80",  label:"Палета" },  
	  { type:"ro", sort:"date", align:"left",  width:"64",  label:"С даты" },
	  
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"ID2" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"test1" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"test2" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"test3" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"test4" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"test5" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"test6" },
	  { type:"ro", sort:"str",  align:"left",  width:"0",   label:"test7" }
  
    ]});

	sw_grid.enableColumnAutoSize(true);
	
						var col_checked = 1;
						var col_block = 2;
						var col_cell = 3;
					

						
                        window.Cleaner.push(sw_grid);
                        
                        
                        /*
                    	sw_grid.attachEvent("onFilterStart",function(){
                    		
                    		//window.update();
                        })
                        */
                       
                       
                        
                        window.update = function(){ 

                        	params = {}
                        	
                        	self._messager.clearAll()
                        	
                        	var is_filled = false; 

                        	for (var i=0; i<15; i++){
                        		
                        		item = sw_grid.getFilterElement(i);
                        		
                        		if (item && item.value){
                        			params['dhx_filter[' + i + ']'] = item.value;
                        			
                        			is_filled = true;
                        		}                        		
                        		
                        	}
                        	
                        	/*
                        	if (!is_filled){
                        		
                        		app.AddMessage('Фильтры для ячеек не выбраны', 2)
                        		return false;
                        	} 
                        	*/
                        
                            params["departs"]  = self.Toolbars["def"].getListOptionSelected("departs");
                        	params["is_all"]   = self.Toolbars["def"].getListOptionSelected("is_all");   
                        	params["is_block"] = self.Toolbars["def"].getListOptionSelected("is_block");
                        	params["godenDo"]  = self.Toolbars["def"].getListOptionSelected("godenDo");
           	
                        	window.sw_grid = sw_grid;
                        	
                        	self.ShowPanel("def");
                            self.load(sw_grid, "/warehouse/wareaddr/data?" + self.serialize(params), function() {
                            	
                            	
                            	window.sw_grid.forEachRow(function(id){

                                 		
                                 	if ( window.sw_grid.cells(id,col_block).getValue() == "-1" ){
                                 		window.sw_grid.setRowTextStyle(id, "background-color: #ffc9da;");
                                 	}

                                      
                                 })

                            });
                        	
                        }
                        
                                                
                        window.do_tool_csv = function(){ self.GridCSV(sw_grid) }
                        
                        
                        var swForm = new dhtmlXForm("sw_form", [            
		    {type:"input",  labelWidth: 100, width:100, name:"addr",   bind:"addr",   labelAlign: "right", label:"Ячейка:"},
		    {type:"input",  labelWidth: 100, width:100, name:"inbox",  bind:"inbox",  labelAlign: "right", label:"В коробке:"},
		    {type:"input",  labelWidth: 100, width:100, name:"box",    bind:"box",    labelAlign: "right", label:"Коробок:",},
		    {type:"input",  labelWidth: 100, width:100, name:"code",   bind:"code",   labelAlign: "right", label:"Код товара:"},
		    {type:"input",  labelWidth: 100, width:100, name:"num",    bind:"num",    labelAlign: "right", label:"Номер партии:"},						    
		    {type: "combo", labelWidth: 100, width:110, name:"condition_code", bind:"condition_code", labelAlign: "right", label:"Статус партии:", options:[
		                                                                 
		                                                                {value: "", text: ""},
																		{% for item in all_status %}
																			{value: "{{ item[0] }}", text: "{{ item[1] }}"},
																		{% end %}

						                                                    	]}
						   ]);                        

                        
                        
                        window.Cleaner.push(swForm);
                    
                        
                        
                        window.do_tool_edit_form = function(ids){
                            swForm.ids=ids;
                            
                            swForm.clear();
                            swForm.lock();
                            
                            
                            dhtmlxAjax.get("/warehouse/wareaddr/data?psw_id="+ids, function(loader) {
                            	
                            	
                            	var all_results =JSON.parse(loader.xmlDoc.responseText);
                            	
                    			for (var item in all_results) {
     
                            		try{
                            			
                            			swForm.setItemValue(item, all_results[item]);
                            			
                            		}
                            		catch(e){
                            			
                            		}
                            	}
                            	
                            
                            	
                            	swForm.unlock();
                            	window.swForm = swForm;

                        	});
                            	
                                                        
                            
                            self.ShowPanel("one");
                    
                            
                                                    
                           
                        }


                        //Правка
                        window.do_tool_edit = function(){
                            var ids = sw_grid.getSelectedRowId();
                            
                            if (!ids){ app.AddMessage('Выберите ячейку',2) }
                            else     { window.do_tool_edit_form(ids) }
                        }
                        
                        
                      
                        window.do_block = function(){
                            var ids = sw_grid.getSelectedRowId();
                            
                            if (!ids){ 
                        			app.AddMessage('Выберите ячейку',2)
                        			return false;
                        	}
                                                                                    
                         	window.sw_grid.setRowTextStyle(ids, "background-color: #ffc9da;");                         	
                            
                         	var cell_id =  window.sw_grid.cells(ids,col_cell).getValue();
                      
                         	dhtmlxAjax.post("/mbl/inventory/block_cell", 'cell_id=' + cell_id  , function(resp){
                         		alert("Ячейка заблокирована");
                         	});
                        }
                        
                        
                        
                        window.do_unblock = function(){
                            var ids = sw_grid.getSelectedRowId();
                            
                            if (!ids){ 
                        			app.AddMessage('Выберите ячейку',2)
                        			return false;
                        	}
                                                                                    
                         	window.sw_grid.setRowTextStyle(ids, "background-color: #fff;");                         	
                            
                         	var cell_id =  window.sw_grid.cells(ids,col_cell).getValue();
                         	 
                         	
                         	dhtmlxAjax.post("/mbl/inventory/unblock_cell", 'cell_id=' + cell_id  , function(resp){
                         		alert("Ячейка разблокирована");
                         	});
                        }
                        
                        
                        function print_format(format){
    
                        	sw_grid.filterBy(0,0);
                        	
                            var ids = sw_grid.getCheckedRows(0);
                            
                            if (!ids){ 
                        			app.AddMessage('Выберите ячейку',2)
                        			return false;
                        	}
                          
                            var all_ids =  ids.split(",");
                                                      
                                                        
                            var sw_grid_print = new dhtmlXGridObject({                                                  
                                columns:[
                                                                                  	
                                         	{ type:"ro", sort:"str",  align:"center",  width:"100", label:"Ячейка" },                                         	
                                         	{ type:"ro", sort:"str",  align:"left",    width:"100", label:"Код Товара" },
                                         	{ type:"ro", sort:"str",  align:"left",    width:"300", label:"Наименование" },
                                         	{ type:"ro", sort:"str",  align:"center",  width:"50",  label:"В коробке" },
                                         	{ type:"ro", sort:"str",  align:"center",  width:"50",  label:"Коробок" },
                                         	{ type:"ro", sort:"str",  align:"center",  width:"50",  label:"Кол-во" },
                                         	{ type:"ro", sort:"str",  align:"center",  width:"100", label:"Партия" },
                                         	{ type:"ro", sort:"str",  align:"center",  width:"100", label:"Годен до" },
                                         	{ type:"ro", sort:"str",  align:"center",  width:"100", label:"С даты" },
                                         	
                                         ]
                            });
        
                    
                            var custom_rows = [3, 6, 7, 8, 9, 10, 11, 12, 15 ];
                                                        
                            for (var i = 0; i < all_ids.length; i++) {
                            	row_id = all_ids[i];

                            	var new_row = [];
                            	
                                for (var j = 0; j < custom_rows.length; j++) {
                                	new_row.push( sw_grid.cellById(row_id, custom_rows[j]).getValue() );
                                }
                                	                                                            	
                            	sw_grid_print.addRow(row_id, new_row.join(','));
                            }
                                     
                            
                            if (format == "print"){
                            	sw_grid_print.printView("","<script> window.print(); </script>");
                            }
                            else if (format == "csv"){
                            	
                            	self.GridCSV(sw_grid_print);
                            }
                            
                             
                            sw_grid_print.destructor();
                        }
                        
                        
                        
                        window.do_csv = function(){
                        	print_format("csv");
                        }
                        
                        window.do_print = function(){
                        	print_format("print");
                        }
                        
        
                        
                        
                        window.do_print_party = function(){
                        	
                        	
                        	var row_id = sw_grid.getSelectedId();
                        	var party  = sw_grid.cellById(row_id, 18).getValue();
                        	
                        	self.NetSend("/btk/btkconfirmin/data/print?party=" + party);
                        }
                        
                        
                        window.do_csv_all = function(){
                        	self.GridCSV(sw_grid);
                        }
                        
                        //Cancel
                        window.do_cancel = function(){ self.ShowPanel("def") }
                        
                        
                        window.do_save = function(){ 	
                        	
                        	//swForm.updateValues();
                        	swForm.validate();
                        	
                        	
                            var dt=swForm.Serialize();
                            if(!!swForm.ids){
                                dt=dt+"&uid="+swForm.ids;
                            }
                            
                            
                            self.NetSendAsync("/warehouse/wareaddr/data", dt, function(){
                            	
                            	window.update();
                            	  
                            });
                                                 
                        }
                        
                        
               
                        //window.update();

File: warehouse/wareaddr.py
# coding: utf-8

from core.sw_base import BaseHandler
from avtuk.avtuk_models import RC


from tornado import template
import cx_Oracle

import os
from settings import ROOT_DIR

from core.common_tools import fetchone, fetchall


#=== Адреса хранения ==========================================================
class WareAddressHandler(BaseHandler):
    urls = r'/warehouse/wareaddr'


    def get(self):


        out = self.cursor.var(cx_Oracle.CURSOR)

        all_tovar_types = self.proc("shiva.GetTovarTypeList", [self.session.rc, out])

        # список типов товара
        tovar_types = [{"id": i[0], "type":"button", "text": i[1]} for i in all_tovar_types[-1].fetchall()]
        
        
        all_storage_types = self.proc("shiva.GetStorageTypeList", [self.session.rc, out])    

        # список типов хранения
        storage_types = [{"id": i[0], "type":"button", "text":i[1]} for i in all_storage_types[-1].fetchall()]                
        
        
        res = self.execute("select id,name from vw_iq_status_cell")
        all_status = res.fetchall()
        
        all_items = [
                            
                           {'type':"button", 'text':'Печать', 'img':"print.gif", 'imgdis':"print_dis.gif", 'action':'do_print'},
                           {'type':"button", 'text':'Печать партии', 'img':"print.gif", 'imgdis':"print_dis.gif", 'action':'do_print_party'},

                           {'type':"button", 'text':'Выделенные в CSV', 'img':"excel24.png", 'imgdis':"excel24g.png", 'action':'do_csv', 'title':'Сохранить в CSV'},
                           {'type':"button", 'text':'Все в CSV', 'img':"excel24.png", 'imgdis':"excel24g.png", 'action':'do_csv_all', 'title':'Сохранить все в CSV'},
                 ]
    

        edit_tools = [
                        {'type':"button", 'text':'Правка', 'img':"edit24.png", 'imgdis':"edit24g.png", 'action':'do_tool_edit'},
                        {'type':"separator"},
                        {'type':"button", 'text':"Заблокировать", 'action':'do_block', 'title':'Заблокировать'},
                        {'type':"button", 'text':"Разблокировать", 'action':'do_unblock', 'title':'Разблокировать'},
                        {'type':"separator"},
                      ]
        
        if self.session.role != "26":
            all_items = edit_tools + all_items
            

          
        all_departs = RC.get_current(rc=self.session.rc).depart_cls
        # кнопки с департаментами РЦ  
        departs = [{"id":"%s" % i.id, "type":"button", "text":i.name}
                   for i in all_departs]


                        
        # фильтр ячеек
        is_all = [
                  {"id":"0", "type":"button", "text":"Все"},
                  {"id":"1", "type":"button", "text":"Заполненные"},                  
                  {"id":"2", "type":"button", "text":"Пустые"},
                 ]


        # фильтр ячеек
        is_block = [{"id":"0", "type":"button", "text":"Все"},
                    {"id":"1", "type":"button", "text":"Не заблокированные"},
                    {"id":"2", "type":"button", "text":"Заблокированные"},
                   ]
                    

        # годен до
        godenDo = [{"id":"0",    "type":"button", "text":"Не выбрано"},
                    {"id":"90",  "type":"button", "text":"Истекает не позже 90 дней"},
                    {"id":"180", "type":"button", "text":"Истекает не позже 180 дней"},
                   ]                    
                    
            
        loader = template.Loader(os.path.join(ROOT_DIR, 'warehouse'))
        result = loader.load("wareaddr.js").generate(departs=departs[0], is_all=is_all[0], is_block=is_block[0], all_status=all_status, godenDo=godenDo[0])            
    
        self.write({'def':[
                           
                                                     
                            {'id':'bdep', 'type':"buttonSelect", 'text':'Режимы', 'items': all_items },
                           
                            {'type':"separator"},
                           
                           
                           {'type':"text", 'text':"Подразделение:"},
                           {'id':'departs', 'type':"buttonSelect",  'mode':'select', 'selected': all_departs[0].id,  'items':departs},
                           {'type':"separator"},

                           {'type':"text", 'text':"Ячейки:"},
                           {'id':'is_all', 'type':"buttonSelect",   'mode':'select', 'selected': '1', 'items':is_all},
                           {'type':"separator"},

                           {'type':"text", 'text':"Блокировки:"},
                           {'id':'is_block', 'type':"buttonSelect", 'mode':'select', 'selected': '0', 'items':is_block},
                           {'type':"separator"},
                           
                           
                           {'type':"text", 'text':"Срок Годности:"},
                           {'id':'godenDo', 'type':"buttonSelect", 'mode':'select', 'selected': '0', 'items':godenDo},
                           {'type':"separator"},
                      
                      
                      
                          # {'type':"text", 'text':"Типы товара:"},
                          # {'id':'is_block', 'type':"buttonSelect", 'mode':'select', 'selected': '0', 'items':tovar_types},
                          # {'type':"separator"},                           
                           
                          # {'type':"text", 'text':"Типы хранения:"},
                          # {'id':'is_block', 'type':"buttonSelect", 'mode':'select', 'selected': '0', 'items':storage_types},
                          # {'type':"separator"},
                      
                           
                           
                           {'type':"separator"},
                           {'type':"button", 'img':"restart.png", 'action':'update', 'text':'Обновить'},
             
                           ],

                    'one':[{'type':"button", 'text':'Сохранить', 'img':"accept24.png", 'imgdis':"accept24g.png", 'action':'do_save'},
                           {'type':"button", 'text':'Отмена', 'img':"delete24.png", 'imgdis':"delete24g.png", 'action':'do_cancel'},
                           '<div id="sw_form" />'],
                    
                    'cmd': result      
             })



#==============================================================================
class WareAddressDataHandler(BaseHandler):
    urls = r'/warehouse/wareaddr/data'

    def get(self): 
        
        pDepart      = self.get_argument("departs",  default=None)
        is_All       = self.get_argument("is_all",   default=None)
        is_Block     = self.get_argument("is_block", default=None)
        psw_id       = self.get_argument("psw_id",   default=None)  # # Одна ячейка для редактировани
        
        pCellName    = self.get_argument("dhx_filter[3]",  default="")
        pStorage     = self.get_argument("dhx_filter[4]",  default="")
        pTypeTovar   = self.get_argument("dhx_filter[5]",  default="")
        pCode        = self.get_argument("dhx_filter[6]",  default="")
        pName        = self.get_argument("dhx_filter[7]",  default="")
        pParty       = self.get_argument("dhx_filter[11]", default="")
        pPartyStatus = self.get_argument("dhx_filter[13]", default="")
        
        godenDo      = self.get_argument("godenDo", default=None)
        
        
        out = self.cursor.var(cx_Oracle.CURSOR)
        
        
        res = self.proc("shiva.GetSaldoList", [pDepart, is_All, is_Block, psw_id, pCellName,
                                                          pStorage, pTypeTovar, pCode, pName, pParty, pPartyStatus, godenDo, out])                
     
        all_results = None
             
        if psw_id:                           
            all_results = fetchone(res[-1])        
        else:
            all_results = fetchall(res[-1], count=1)
                    
        self.write(all_results)         



    
    def post(self):
        
        sw_id  = self.get_argument("uid",    default=None)
        addr   = self.get_argument("addr",  default=None)
        inbox  = self.get_argument("inbox", default=None)
        box    = self.get_argument("box",   default=None) 
        code   = self.get_argument("code",  default=None)
        num    = self.get_argument("num",   default=None)
        
        condition_code = self.get_argument("condition_code", default=None)

        
        result = self.proc("shiva.EditPalletPartyCell", [sw_id, addr, box, inbox, code, num, condition_code])
        
        self.write({"status": result})
